prover/src/constraints/evaluator.rs

```diff
@@ -335,4 +335,37 @@ impl<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> ConstraintEvaluator<
     fn num_aux_transition_constraints(&self) -> usize {
         self.transition_constraints.num_aux_constraints()
     }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    // This is a modified version of the TransitionConstraintGroup::merge_evaluations method. This
+    // implementation uses pre-computed of powers of generator of the constraint evaluation domain.
+    // This allows up to substitute the exponentiations needed for the degree adjustment factor 
+    // with lookups. 
+    pub fn merge_evaluations_optimized<F>(
+        group: &TransitionConstraintGroup<E>,
+        evaluations: &[F],
+        step: usize,
+        domain: &StarkDomain<A::BaseField>,
+    ) -> E
+    where
+        E: FieldElement<BaseField = A::BaseField>,
+        F: FieldElement<BaseField = A::BaseField> + ExtensionOf<A::BaseField>,
+        E: FieldElement<BaseField = A::BaseField> + ExtensionOf<A::BaseField> + ExtensionOf<F>,
+    {
+
+        let index: usize = step * (group.degree_adjustment as usize);
+        let index = index % (domain.ce_domain_size());
+        let xp = domain.domain_g[index] * *domain.adj_map.get(&group.degree_adjustment).unwrap(); 
```

Comment:
I would probably move this logic into the `StarkDomain` struct. For example, we could add the following function there:
```Rust
pub fn get_ce_x_power_at(&self, step: usize, degree_adjustment: usize) -> A::BaseElement
```